pkg/paravirt/paravirt_protected_entity.go

```diff
@@ -0,0 +1,150 @@
+	"context"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware-tanzu/astrolabe/pkg/astrolabe"
+	"github.com/vmware-tanzu/astrolabe/pkg/pvc"
+	backupdriverv1api "github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/apis/backupdriver/v1"
+	"github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotUtils"
+	"io"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+/*
+ParaVirtProtectedEntity implements a Protected Entity interface to Paravirtualized objects,
+including, Persistent Volumes, VMs, Persistence Services, etc., in the Paravirtualized model
+in vSphere Project Pacific. Paravirtualized Persistent Volume would be the first use case,
+but the ParaVirtProtectedEntity would be implemented to be extensible for other objects.
+
+ParaVirtProtectedEntity has no component PEs as IVDProtectedEntity does. Specifically, the
+whole Paravirtualized Backup Driver stack will serve as the backend of ParaVirtProtectedEntity.
+*/
+
+type ParaVirtProtectedEntity struct {
+	pvpetm   *ParaVirtProtectedEntityTypeManager
+	id       astrolabe.ProtectedEntityID
+	data     []astrolabe.DataTransport
+	metadata []astrolabe.DataTransport
+	combined []astrolabe.DataTransport
+	logger   logrus.FieldLogger
+}
+
+// Object methods
+
+func (this ParaVirtProtectedEntity) GetInfo(ctx context.Context) (astrolabe.ProtectedEntityInfo, error) {
+	var name string
+	var err error
+	if this.pvpetm.entityType == ParaVirtEntityTypePersistentVolume {
+		name, err = this.getVolumeHandleFromPV()
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+	}
+
+	return astrolabe.NewProtectedEntityInfo(this.id, name, this.data, this.metadata, this.combined, []astrolabe.ProtectedEntityID{}), nil
+}
+
+func (this ParaVirtProtectedEntity) GetCombinedInfo(ctx context.Context) ([]astrolabe.ProtectedEntityInfo, error) {
+	paravirtIPE, err := this.GetInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return []astrolabe.ProtectedEntityInfo{paravirtIPE}, nil
+}
+
+func (this ParaVirtProtectedEntity) Snapshot(ctx context.Context, params map[string]map[string]interface{}) (astrolabe.ProtectedEntitySnapshotID, error) {
+	this.logger.Infof("CreateSnapshot called on Paravirtualized Protected Entity, %v", this.id.String())
+	// create snapshot CR using snapshot utils
+	peInfo, err := this.GetInfo(context.TODO())
+	if err != nil {
+		this.logger.Errorf("Failed to get info for ParaVirtProtectedEntity %v", this.id.String())
+		return astrolabe.ProtectedEntitySnapshotID{}, errors.WithStack(err)
+	}
+	objectToSnapshot := corev1.TypedLocalObjectReference{
+		APIGroup: &corev1.SchemeGroupVersion.Group,
+		Kind:     "PersistentVolumeClaim", // PVC kind
+		Name:     peInfo.GetName(),        // Supervisor PVC Name, i.e., Guest PV CSI VolumeHandle
+	}
+
+	// TODO: the key of params might get changed
+	backupRepositoryName, ok := params[astrolabe.PvcPEType]["BackupRepositoryName"].(string)
+	if !ok {
+		backupRepositoryName = "INVALID_BR_NAME"
```

Comment:
Should we leave the string empty? In local mode, maybe it will be empty? We can make this change later as well.

Reply:
Sure. It is just a place holder for the moment. We refine it when work on the data path.